pkg/ffmpeg/stream.go

```diff
@@ -0,0 +1,244 @@
+package ffmpeg
+
+import (
+	"io"
+	"io/ioutil"
+	"net/http"
+	"os"
+	"os/exec"
+	"strconv"
+	"strings"
+
+	"github.com/stashapp/stash/pkg/logger"
+	"github.com/stashapp/stash/pkg/models"
+)
+
+const CopyStreamCodec = "copy"
+
+type Stream struct {
+	Stdout   io.ReadCloser
+	Process  *os.Process
+	options  TranscodeStreamOptions
+	mimeType string
+}
+
+func (s *Stream) Serve(w http.ResponseWriter, r *http.Request) {
+	w.Header().Set("Content-Type", s.mimeType)
+	w.WriteHeader(http.StatusOK)
+
+	logger.Infof("[stream] transcoding video file to %s", s.mimeType)
+
+	// handle if client closes the connection
+	notify := r.Context().Done()
+	go func() {
+		<-notify
+		s.Process.Kill()
+	}()
+
+	_, err := io.Copy(w, s.Stdout)
+	if err != nil {
+		logger.Errorf("[stream] error serving transcoded video file: %s", err.Error())
+	}
+}
+
+type Codec struct {
+	Codec     string
+	format    string
+	MimeType  string
+	extraArgs []string
+	hls       bool
+}
+
+var CodecHLS = Codec{
+	Codec:    "libx264",
+	format:   "mpegts",
+	MimeType: MimeMpegts,
+	extraArgs: []string{
+		"-acodec", "aac",
+		"-pix_fmt", "yuv420p",
+		"-preset", "veryfast",
+		"-crf", "25",
+	},
+	hls: true,
+}
+
+var CodecH264 = Codec{
+	Codec:    "libx264",
+	format:   "mp4",
+	MimeType: MimeMp4,
+	extraArgs: []string{
+		"-movflags", "frag_keyframe",
+		"-pix_fmt", "yuv420p",
+		"-preset", "veryfast",
+		"-crf", "25",
+	},
+}
+
+var CodecVP9 = Codec{
+	Codec:    "libvpx-vp9",
+	format:   "webm",
+	MimeType: MimeWebm,
+	extraArgs: []string{
+		"-deadline", "realtime",
+		"-cpu-used", "5",
+		"-row-mt", "1",
+		"-crf", "30",
+		"-b:v", "0",
+	},
+}
+
+var CodecVP8 = Codec{
```

Comment:
given that only one webm endpoint is available VP9 should be the preferred codec
thus VP8 settings can be removed altogether.
If we choose to keep them they should be tuned a little bit 
From some tests to get acceptable quality i had to alter the `crf` to `12`, `-b:v` to `3M` ( 0 doesn't seem to work like it did for VP9) and add an extra `"-pix_fmt", "yuv420p",`

Reply:
I've left the VP8 codec in there for now, with your revised settings.